### ISSUE TITLE
ChaCha priority detection logic broken

### ISSUE THREAD

**Issue by GermanCoding** [syncthing/syncthing#7593]
### Does your log mention database corruption?

No.

### Include required information

Syncthing v1.15.1 "Fermium Flea" (go1.16.3 *-amd64) teamcity@build.syncthing.net 2021-04-06 08:42:29 UTC

## What happened

Syncthing uses PreferServerCipherSuites: true on TLS connections, but uses custom detection (by sneakily asking the Go runtime) to determine whether ChaCha should be used, or if AES-GCM should be preferred. This used to work just fine. However, in recent versions of syncthing, **all** of my machines now use TLS1.3-TLS_CHACHA20_POLY1305_SHA256. Those machines previously always used TLS1.3-TLS_AES_128_GCM_SHA256. All machines involved have AES hardware support (AES-NI).

I think this may be due to a change in Go's runtime, as the [Go 1.16 changelog](https://golang.org/doc/go1.16)* mentions that something in this regard has changed.

I believe syncthing's manual interventions on cipher preference somehow intervenes with the ChaCha/AES-GCM detection logic, which always causes ChaCha to be selected. This isn't a pressing issue, but on my machines AES-GCM definetly uses less CPU, as it's mostly in hardware, while ChaCha is not.

Note: This broke recently. I know that Go's handling of cipher suites has changed quite a lot with the introduction of TLS 1.3, but this broke much later - I have only done very limited testing, but it is likely related to Go 1.16.

> *Servers will now prefer other available AEAD cipher suites (such as ChaCha20Poly1305) over AES-GCM cipher suites if either the client or server doesn't have AES hardware support, unless both Config.PreferServerCipherSuites and Config.CipherSuites are set. The client is assumed not to have AES hardware support if it does not signal a preference for AES-GCM cipher suites.

**Comment by calmh** [comment]
You show TLS 1.3 connections, and for those we don't control the cipher suites at all - the Go standard library does what it wants. Something might have changed in Go 1.16 though.

**Comment by GermanCoding** [comment]
> You show TLS 1.3 connections, and for those we don't control the cipher suites at all

That is correct, but you can still influence the priority of AES-GCM in TLS 1.3 connections :) - it's completly weird in 1.16. I think I have figured out what happens:

Syncthing determines the order of TLS1.2-CipherSuites. This either puts ChaCha or AES-GCM on the top. Right now this is seemingly broken and always puts ChaCha on the top for TLS 1.2 connections. Let's have a look on how this looks in practice:

![syncthing_tls_cipher_order_edited](https://user-images.githubusercontent.com/4279661/115867528-823d1900-a43b-11eb-8c33-b9ec249bb232.png)

This shows the cipher preference of a **syncthing client**. As we can see here, the TLS1.2 ciphers always go first. Syncthing **does** influence the ordering of TLS1.2 suites.

Now, what happens? The new Go 1.6 logic **on the server** determines whether the **very, very first cipher suite** is AES-GCM. It is not, as syncthing has set ChaCha as most preferred. [The Go server will thus select some TLS1.3 ChaCha suite](https://golang.org/src/crypto/tls/handshake_server_tls13.go), as that is what the go server deduces it what the client wants. (It doesn't matter that the very first cipher suite is not a TLS 1.3 suite)

**TL;DR** Syncthings (incorrect) ordering of TLS 1.2 suites confused the selection logic on the server, always selecting ChaCha.

Why is Syncthings ordering of TLS 1.2 suites incorrect? It used to work previously.

-> I'm not sure yet, probably the change in selection logic on Go 1.16 broke it. Still investigating.

**Comment by GermanCoding** [comment]
This was effectively solved by https://github.com/syncthing/syncthing/pull/7598:

With the default settings - InsecureAllowOldTLSVersions set to off - the Go library will select cipher ordering*, thus prioritizing either AES-GCM or ChaCha as before.

*On a side note, Go still sends TLS 1.2 cipher suites even with TLS 1.2 turned off 🤔, but anyway, as the order of those is determined by the stdlib, it works.

With non-default settings - InsecureAllowOldTLSVersions set to on - the order is static, always preferring ChaCha (both on TLS 1.3 and TLS 1.2). So now it is not a bug, it's a feature 👍. I think that's an acceptable side effect for running with legacy settings, as that's phase out anyway.
